server/src/domain/database/database.service.spec.ts

```diff
@@ -0,0 +1,134 @@
+      expect(databaseMock.runMigrations).toHaveBeenCalledTimes(1);
+      expect(fatalLog).not.toHaveBeenCalled();
+    });
+
+    it('should throw an error if vectors version is not installed even after createVectors', async () => {
+      databaseMock.getExtensionVersion.mockResolvedValueOnce(null);
+
+      await expect(sut.init()).rejects.toThrow('Unexpected: The pgvecto.rs extension is not installed.');
+
+      expect(databaseMock.getExtensionVersion).toHaveBeenCalledTimes(1);
+      expect(databaseMock.createExtension).toHaveBeenCalledTimes(1);
+      expect(databaseMock.runMigrations).not.toHaveBeenCalled();
+    });
+
+    it('should throw an error if vectors version is below minimum supported version', async () => {
+      databaseMock.getExtensionVersion.mockResolvedValueOnce(new Version(0, 0, 1));
+
+      await expect(sut.init()).rejects.toThrow(/('tensorchord\/pgvecto-rs:pg14-v0.1.11')/s);
+
+      expect(databaseMock.getExtensionVersion).toHaveBeenCalledTimes(1);
+      expect(databaseMock.runMigrations).not.toHaveBeenCalled();
+    });
+
+    it('should throw an error if vectors version is above maximum supported version', async () => {
+      databaseMock.getExtensionVersion.mockResolvedValueOnce(new Version(0, 1, 12));
+
+      await expect(sut.init()).rejects.toThrow(
+        /('DROP EXTENSION IF EXISTS vectors').*('tensorchord\/pgvecto-rs:pg14-v0\.1\.11')/s,
+      );
+
+      expect(databaseMock.getExtensionVersion).toHaveBeenCalledTimes(1);
+      expect(databaseMock.runMigrations).not.toHaveBeenCalled();
+    });
+
+    it('should throw an error if vectors version is a nightly', async () => {
+      databaseMock.getExtensionVersion.mockResolvedValueOnce(new Version(0, 0, 0));
+
+      await expect(sut.init()).rejects.toThrow(
+        /(nightly).*('DROP EXTENSION IF EXISTS vectors').*('tensorchord\/pgvecto-rs:pg14-v0\.1\.11')/s,
+      );
+
+      expect(databaseMock.getExtensionVersion).toHaveBeenCalledTimes(1);
+      expect(databaseMock.createExtension).toHaveBeenCalledTimes(1);
+      expect(databaseMock.runMigrations).not.toHaveBeenCalled();
+    });
+
+    it('should throw error if vectors extension could not be created', async () => {
+      databaseMock.createExtension.mockRejectedValueOnce(new Error('Failed to create extension'));
+
+      await expect(sut.init()).rejects.toThrow('Failed to create extension');
+
+      expect(fatalLog).toHaveBeenCalledTimes(1);
+      expect(fatalLog.mock.calls[0][0]).toMatch(/('tensorchord\/pgvecto-rs:pg14-v0\.1\.11').*(v1\.91\.0)/s);
+      expect(databaseMock.createExtension).toHaveBeenCalledTimes(1);
+      expect(databaseMock.runMigrations).not.toHaveBeenCalled();
+    });
+
+    it('should set the image to the correct postgres version', async () => {
+      databaseMock.getExtensionVersion.mockResolvedValue(new Version(0, 0, 1));
+      [14, 15, 16].forEach((major) => databaseMock.getPostgresVersion.mockResolvedValueOnce(new Version(major, 0, 0)));
```

Comment:
You could put the for loop outside the it

Reply:
Ooh, good idea